File: qeli-shared/QeliShared/Geo/GeoAssetStore.cs
namespace Qeli.Shared.Geo;

/// <summary>Download + cache directory for geosite.dat / geoip.dat.</summary>
public static class GeoAssetStore
{
    public const string GeositeFile = "geosite.dat";
    public const string GeoipFile = "geoip.dat";

    // Primary: RU-oriented lists (same source v2rayN RUv1 presets use).
    // Fallback: Loyalsoldier community dat.
    // Prefer CDN/raw first — github.com/releases often hits HTTP/2 REFUSED_STREAM behind proxies.
    public static readonly string[] GeositeUrls =
    {
        "https://cdn.jsdelivr.net/gh/runetfreedom/russia-v2ray-rules-dat@release/geosite.dat",
        "https://raw.githubusercontent.com/runetfreedom/russia-v2ray-rules-dat/release/geosite.dat",
        "https://github.com/runetfreedom/russia-v2ray-rules-dat/releases/latest/download/geosite.dat",
        "https://github.com/Loyalsoldier/v2ray-rules-dat/releases/latest/download/geosite.dat",
    };

    public static readonly string[] GeoipUrls =
    {
        "https://cdn.jsdelivr.net/gh/runetfreedom/russia-v2ray-rules-dat@release/geoip.dat",
        "https://raw.githubusercontent.com/runetfreedom/russia-v2ray-rules-dat/release/geoip.dat",
        "https://github.com/runetfreedom/russia-v2ray-rules-dat/releases/latest/download/geoip.dat",
        "https://github.com/Loyalsoldier/v2ray-rules-dat/releases/latest/download/geoip.dat",
    };

    private static string _dir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "Qeli", "geo");

    private static readonly object Gate = new();
    private static GeoSiteIndex? _site;
    private static GeoIpIndex? _ip;
    private static string _loadedPreset = "";

    public static string DirectoryPath => _dir;
    public static string GeositePath => Path.Combine(_dir, GeositeFile);
    public static string GeoipPath => Path.Combine(_dir, GeoipFile);

    public static bool HasFiles => File.Exists(GeositePath) && File.Exists(GeoipPath);

    public static string StatusText()
    {
        if (!HasFiles) return "not downloaded";
        var gs = new FileInfo(GeositePath);
        var gi = new FileInfo(GeoipPath);
        return $"geosite {gs.Length / 1024} KB, geoip {gi.Length / 1024} KB — {gs.LastWriteTime:yyyy-MM-dd}";
    }

    public static void Configure(string directory)
    {
        _dir = directory;
        lock (Gate)
        {
            _site = null;
            _ip = null;
            _loadedPreset = "";
        }
    }

    public static async Task DownloadAsync(Action<string>? log = null, CancellationToken ct = default)
    {
        Directory.CreateDirectory(_dir);
        await DownloadOneAsync(GeositeUrls, GeositePath, "geosite.dat", log, ct).ConfigureAwait(false);
        await DownloadOneAsync(GeoipUrls, GeoipPath, "geoip.dat", log, ct).ConfigureAwait(false);
        lock (Gate)
        {
            _site = null;
            _ip = null;
            _loadedPreset = "";
        }
    }

    private static async Task DownloadOneAsync(
        string[] urls, string dest, string label, Action<string>? log, CancellationToken ct)
    {
        Exception? last = null;
        foreach (var url in urls)
        {
            try
            {
                log?.Invoke($"Downloading {label}…");
                using var http = new HttpClient { Timeout = TimeSpan.FromMinutes(5) };
                http.DefaultRequestHeaders.UserAgent.ParseAdd("qeli-geo/0.7");
                http.DefaultRequestVersion = System.Net.HttpVersion.Version11;
                http.DefaultVersionPolicy = System.Net.Http.HttpVersionPolicy.RequestVersionOrLower;
                using var resp = await http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead, ct)
                    .ConfigureAwait(false);
                resp.EnsureSuccessStatusCode();
                var tmp = dest + ".tmp";
                await using (var fs = File.Create(tmp))
                {
                    await resp.Content.CopyToAsync(fs, ct).ConfigureAwait(false);
                }
                if (new FileInfo(tmp).Length < 1024)
                    throw new InvalidDataException($"{label} too small");
                File.Copy(tmp, dest, overwrite: true);
                File.Delete(tmp);
                log?.Invoke($"{label} OK ({new FileInfo(dest).Length / 1024} KB)");
                return;
            }
            catch (Exception e)
            {
                last = e;
                log?.Invoke($"{label} failed from {url}: {e.Message}");
            }
        }
        throw last ?? new IOException($"could not download {label}");
    }

    public static (GeoSiteIndex? Site, GeoIpIndex? Ip) GetOrLoad(string presetId)
    {
        var need = ProxyRoutePreset.TagsFor(presetId);
        lock (Gate)
        {
            if (_site != null && _ip != null && _loadedPreset == presetId)
                return (_site, _ip);
            if (!HasFiles) return (null, null);
            _site = GeoSiteIndex.Load(GeositePath, need.SiteTags);
            _ip = GeoIpIndex.Load(GeoipPath, need.IpTags);
            _loadedPreset = presetId;
            return (_site, _ip);
        }
    }

    public static void Invalidate()
    {
        lock (Gate)
        {
            _site = null;
            _ip = null;
            _loadedPreset = "";
        }
    }
}

File: qeli-android/app/src/main/kotlin/com/qeli/geo/GeoAssetStore.kt
package com.qeli.geo

import android.content.Context
import android.os.Build
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

/** Download + cache geosite.dat / geoip.dat (same sources as Windows GeoAssetStore). */
object GeoAssetStore {
    const val GEOSITE_FILE = "geosite.dat"
    const val GEOIP_FILE = "geoip.dat"

    private val geositeUrls = listOf(
        "https://cdn.jsdelivr.net/gh/runetfreedom/russia-v2ray-rules-dat@release/geosite.dat",
        "https://raw.githubusercontent.com/runetfreedom/russia-v2ray-rules-dat/release/geosite.dat",
        "https://github.com/runetfreedom/russia-v2ray-rules-dat/releases/latest/download/geosite.dat",
        "https://github.com/Loyalsoldier/v2ray-rules-dat/releases/latest/download/geosite.dat",
    )
    private val geoipUrls = listOf(
        "https://cdn.jsdelivr.net/gh/runetfreedom/russia-v2ray-rules-dat@release/geoip.dat",
        "https://raw.githubusercontent.com/runetfreedom/russia-v2ray-rules-dat/release/geoip.dat",
        "https://github.com/runetfreedom/russia-v2ray-rules-dat/releases/latest/download/geoip.dat",
        "https://github.com/Loyalsoldier/v2ray-rules-dat/releases/latest/download/geoip.dat",
    )

    @Volatile private var site: GeoSiteIndex? = null
    @Volatile private var ip: GeoIpIndex? = null
    @Volatile private var loadedPreset = ""

    fun dir(ctx: Context): File = File(ctx.filesDir, "geo").also { it.mkdirs() }
    fun geositePath(ctx: Context) = File(dir(ctx), GEOSITE_FILE)
    fun geoipPath(ctx: Context) = File(dir(ctx), GEOIP_FILE)
    fun hasFiles(ctx: Context) = geositePath(ctx).isFile && geoipPath(ctx).isFile

    fun statusText(ctx: Context): String {
        if (!hasFiles(ctx)) return "not downloaded"
        val gs = geositePath(ctx)
        val gi = geoipPath(ctx)
        return "geosite ${gs.length() / 1024} KB, geoip ${gi.length() / 1024} KB"
    }

    suspend fun download(ctx: Context, log: (String) -> Unit = {}) = withContext(Dispatchers.IO) {
        dir(ctx)
        downloadOne(geositeUrls, geositePath(ctx), "geosite.dat", log)
        downloadOne(geoipUrls, geoipPath(ctx), "geoip.dat", log)
        site = null
        ip = null
        loadedPreset = ""
    }

    fun getOrLoad(ctx: Context, presetId: String): Pair<GeoSiteIndex?, GeoIpIndex?> {
        val id = ProxyRoutePreset.normalize(presetId)
        if (site != null && ip != null && loadedPreset == id) return site to ip
        if (!hasFiles(ctx)) return null to null
        val (siteTags, ipTags) = ProxyRoutePreset.tagsFor(id)
        site = if (siteTags.isEmpty()) GeoSiteIndex() else
            runCatching { GeoSiteIndex.load(geositePath(ctx), siteTags) }.getOrNull()
        ip = if (ipTags.isEmpty()) GeoIpIndex() else
            runCatching { GeoIpIndex.load(geoipPath(ctx), ipTags) }.getOrNull()
        loadedPreset = id
        return site to ip
    }

    /**
     * CIDRs to exclude from TUN for bypass presets.
     * Aggregates geoip blocks and caps the route table on older Android builds.
     */
    fun tunExcludeCidrs(ctx: Context, presetId: String, maxRoutes: Int = tunExcludeRouteBudget()): List<String> {
        val codes = ProxyRoutePreset.tunBypassIpCodes(presetId)
        if (codes.isEmpty()) return emptyList()
        val (_, ipIdx) = getOrLoad(ctx, presetId)
        val raw = ipIdx?.allCidrsFor(codes) ?: return emptyList()
        if (raw.isEmpty()) return emptyList()
        return CidrAggregator.prioritizeForCoverage(raw, maxRoutes)
    }

    /** CIDRs routed into the tunnel for proxy-only presets (ru-blocked, gfw-blacklist). */
    fun tunIncludeCidrs(ctx: Context, presetId: String, maxRoutes: Int = tunIncludeRouteBudget()): List<String> {
        val codes = ProxyRoutePreset.tunIncludeIpCodes(presetId)
        if (codes.isEmpty()) return emptyList()
        val (_, ipIdx) = getOrLoad(ctx, presetId)
        val raw = ipIdx?.allCidrsFor(codes) ?: return emptyList()
        if (raw.isEmpty()) return emptyList()
        return CidrAggregator.prioritizeForCoverage(raw, maxRoutes)
    }

    fun tunIncludeRouteBudget(): Int =
        if (Build.VERSION.SDK_INT >= 33) 2048 else 120

    /** Route budget for geo excludes; pre-Android-13 complement splits are much tighter. */
    fun tunExcludeRouteBudget(): Int =
        if (Build.VERSION.SDK_INT >= 33) 4096 else 160

    private fun downloadOne(urls: List<String>, dest: File, label: String, log: (String) -> Unit) {
        var last: Exception? = null
        for (url in urls) {
            try {
                log("Downloading $label…")
                val conn = (URL(url).openConnection() as HttpURLConnection).apply {
                    connectTimeout = 30_000
                    readTimeout = 120_000
                    setRequestProperty("User-Agent", "qeli-geo/0.7")
                    instanceFollowRedirects = true
                }
                conn.inputStream.use { input ->
                    val tmp = File(dest.absolutePath + ".tmp")
                    tmp.outputStream().use { input.copyTo(it) }
                    if (tmp.length() < 1024) throw IllegalStateException("$label too small")
                    tmp.copyTo(dest, overwrite = true)
                    tmp.delete()
                }
                conn.disconnect()
                log("$label OK (${dest.length() / 1024} KB)")
                return
            } catch (e: Exception) {
                last = e
                log("$label fail (${e.message}) — try next mirror")
            }
        }
        throw last ?: IllegalStateException("download $label failed")
    }
}
